src/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.Windows.cs

```diff
@@ -0,0 +1,73 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+namespace System.Threading
+{
+    /// <summary>
+    /// A LIFO semaphore implemented using Win32 IO Completion Ports.
+    /// </summary>
+    /// <remarks>
+    /// IO Completion ports release waiting threads in LIFO order, so we can use them to create a LIFO semaphore.
+    /// See https://msdn.microsoft.com/en-us/library/windows/desktop/aa365198(v=vs.85).aspx under How I/O Completion Ports Work.
+    /// From the docs "Threads that block their execution on an I/O completion port are released in last-in-first-out (LIFO) order."
+    /// </remarks>
+    internal sealed class LowLevelLifoSemaphore : IDisposable
+    {
+        private IntPtr _completionPort;
+
+        public LowLevelLifoSemaphore(int initialSignalCount, int maximumSignalCount)
+        {
+            Debug.Assert(initialSignalCount >= 0, "Windows LowLevelLifoSemaphore does not support a negative signal count"); // TODO: Track actual signal count to enable this
+            _completionPort = Interop.Kernel32.CreateIoCompletionPort(new IntPtr(-1), IntPtr.Zero, UIntPtr.Zero, 1);
+            if (_completionPort == IntPtr.Zero)
+            {
+                var error = Marshal.GetLastWin32Error();
+                var exception = new OutOfMemoryException();
+                exception.SetErrorCode(error);
+                throw exception;
+            }
+            Release(initialSignalCount);
+        }
+
+        public bool Wait(int timeoutMs)
+        {
+            bool success = Interop.Kernel32.GetQueuedCompletionStatus(_completionPort, out var numberOfBytes, out var completionKey, out var pointerToOverlapped, timeoutMs);
+            if (!success)
+            {
+                var error = Marshal.GetLastWin32Error();
+                Debug.Assert(error != Interop.Kernel32.ERROR_ABANDONDED_WAIT_0, "LowLevelLifoSemaphore waited on after dispose.");
+                if (error != 0)
+                {
+                    var exception = new OutOfMemoryException();
+                    exception.SetErrorCode(error);
+                    throw exception;
```

Comment:
I didn't find it in the documentation but upon timeout, CoreCLR is checking the return value from GetLastError() with WAIT_TIMEOUT to check for a timeout, and that seems to be how it works. Since we don't expect any error codes here other than WAIT_TIMEOUT, could assert that the error code is WAIT_TIMEOUT.